Fix virtual machine hangs after 2x suspend vm breakpoint

raghucssit opened the pull request:
We need to initiate equal number of resume for suspend vm requests by threads.

Fixes https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/320

raghucssit commented:
Changed author as per my eclipse account.
Rebased with master.

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:
```diff
@@ -334,6 +334,10 @@ public class JDIDebugTarget extends JDIDebugElement implements
 	 */
 	private final Map<Long, String> objectLabels = new HashMap<>();
 
+	private int suspendCount;
```
Debugger is always multi-threaded, so all variables should be volatile

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:
```diff
@@ -1340,6 +1344,13 @@ protected void resume(boolean fireNotification) throws DebugException {
 			VirtualMachine vm = getVM();
 			if (vm != null) {
+				if (pendingResumeRequests) {
+					while (suspendCount > 1) {
```
`suspendCount`, once set from zero to `> 1`, will stay always `>= 1` after this while loop.
Why not check `while (suspendCount > 0)`?

iloveeclipse replied:
I've fixed that in my last pushed commit.

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIDebugTarget.java:
```diff
@@ -1340,6 +1344,13 @@ protected void resume(boolean fireNotification) throws DebugException {
+						suspendCount--;
+						pendingResumeRequests = false;
+					}
+				}
```
Shouldn't `vm.resume();` call below be in the `else` block?

raghucssit replied:
This is a default `vm.resume();` which was existing.
When the `suspendCount  = 1` this works by default.
If there were more than 1 Suspend VM breakpoints are set they will be resumed inside while loop.
Whenever a `IJavaBreakpoint.SUSPEND_VM` breakpoint is hit `JDIDebugTarget#suspendedByBreakpoint(JavaBreakpoint breakpoint, boolean queueEvent, EventSet set)` is invoked.
So I am counting the number of Suspend VM breakpoints hit. And whenever VM(Debug Target) Resume is requested all the pending VM suspend will be cleared inside while loop.

iloveeclipse reviewed (changes requested):
Hard to play with / verify the patch without test case.

Please add a test case to `org.eclipse.jdt.debug.tests.breakpoints.SuspendVMBreakpointsTests`

raghucssit commented:
> @raghucssit : I've reworked PR a bit, but the main idea is same & original fix was in right place, so main idea untouched.
> 
> * get rid of the extra boolean flag
> * renamed counter and changed it to a thread safe AtomicInteger
> * added comments in code
> * slightly changed test
> 
> Please review.

Code changes looks good.
I have pulled and did some manual testing, Everything works as expected.
Good to merge.
